source/GameInterface/Services/Towns/Handlers/TownAuditorHandler.cs

```diff
@@ -0,0 +1,152 @@
+using Common.Extensions;
+using Common.Logging;
+using Common.Messaging;
+using GameInterface.Services.ObjectManager;
+using GameInterface.Services.Towns.Data;
+using GameInterface.Services.Towns.Messages;
+using GameInterface.Services.Towns.Patches;
+using Serilog;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using TaleWorlds.CampaignSystem;
+using TaleWorlds.CampaignSystem.Settlements;
+using TaleWorlds.Core;
+using TaleWorlds.Library;
+
+namespace GameInterface.Services.Towns.Handlers
+{
+    /// <summary>
+    /// Handles Town Auditor (send all sync).
+    /// </summary>
+    public class TownAuditorHandler : IHandler
+    {
+        private static readonly ILogger Logger = LogManager.GetLogger<TownAuditorHandler>();
+
+        private readonly IMessageBroker messageBroker;
+        private readonly IObjectManager objectManager;
+        private static readonly Func<Town, Town.SellLog[]> getSoldItems = typeof(Town).GetField("_soldItems", BindingFlags.Instance | BindingFlags.NonPublic).BuildUntypedGetter<Town, Town.SellLog[]>();
+        public TownAuditorHandler(IMessageBroker messageBroker, IObjectManager objectManager)
+        {
+            this.messageBroker = messageBroker;
+            this.objectManager = objectManager;
+
+            messageBroker.Subscribe<SendTownAuditor>(HandleSendTownAuditor);
+        }
+
+        private void HandleSendTownAuditor(MessagePayload<SendTownAuditor> payload)
+        {
+            var obj = payload.What;
+            var clientTownAuditorDatas = obj.Datas;
+
+            List<Settlement> settlements = Campaign.Current.CampaignObjectManager.Settlements
+            .Where(settlement => settlement.IsTown).ToList();
+
+            List<TownAuditorData> serverTownAuditorDatas = new List<TownAuditorData>();
+            settlements.ForEach((settlement) =>
+            {
+
+                if (objectManager.TryGetObject(settlement.Town.StringId, out Town t) == false)
+                {
+                    Logger.Error("TownAuditorHandler: Town not found on server: " + settlement.Town.StringId);
+                }
+                else
+                {
+                    Fief fief = t.Settlement.SettlementComponent as Fief;
+
+                    TownAuditorData auditorData = new TownAuditorData(
+                        t.StringId, t.Name.ToString(), (t.Governor != null) ? t.Governor.Name.ToString() : "null",
+                        (t.LastCapturedBy != null) ? t.LastCapturedBy.Name.ToString() : "null",
+                        t.Prosperity, t.Loyalty, t.Security, t.InRebelliousState, t.GarrisonAutoRecruitmentIsEnabled,
+                       fief.FoodStocks, t.TradeTaxAccumulated, getSoldItems(t));
+
+                    serverTownAuditorDatas.Add(auditorData);
+                }
+
+
+            });
```

Comment:
Move this to a new function

Have a new line for every parameter, currently it is hard to read. I would prefer if you specify the parameter using `paramater_name: value`

"t" is too generic, change to "town"

Finally, ToList is not needed and slower

```suggestion
            IEnumerable<Settlement> settlements = Campaign.Current.CampaignObjectManager.Settlements
            .Where(settlement => settlement.IsTown)
            .Select(settlement =>
            {

                if (objectManager.TryGetObject(settlement.Town.StringId, out Town t) == false)
                {
                    Logger.Error("TownAuditorHandler: Town not found on server: " + settlement.Town.StringId);
                }
                else
                {
                    Fief fief = t.Settlement.SettlementComponent as Fief;

                    return new TownAuditorData(
                        t.StringId,
                        t.Name.ToString(),
                        t.Governor?.Name.ToString() ?? "null",
                        t.LastCapturedBy?.Name.ToString() ?? "null",
                        t.Prosperity,
                        t.Loyalty,
                        t.Security,
                        t.InRebelliousState,
                        t.GarrisonAutoRecruitmentIsEnabled,
                       fief.FoodStocks,
                       t.TradeTaxAccumulated,
                       getSoldItems(t));
                }


            });
```